lib/manageiq/providers/openstack/legacy/events/openstack_stf_event_monitor.rb

```diff
@@ -0,0 +1,112 @@
+require 'manageiq/providers/openstack/legacy/openstack_event_monitor'
+require 'manageiq/providers/openstack/legacy/events/openstack_event'
+require 'manageiq/providers/openstack/legacy/events/openstack_stf_event_converter'
+require 'manageiq/providers/openstack/legacy/events/openstack_stf_event_receiver'
+require 'manageiq/providers/openstack/legacy/events/openstack_stf_event_test_receiver'
+require 'qpid_proton'
+
+class OpenstackStfEventMonitor < OpenstackEventMonitor
+  DEFAULT_AMQP_PORT  = 5666
+  DEFAULT_TOPIC_NAME = 'anycast/ceilometer/event.sample'.freeze
+
+  def self.available?(options = {})
+    $log.info("Testing connection to STF..") if $log
+    $log.debug("With STF options: #{options.inspect}") if $log
+    qdr_client = Qpid::Proton::Container.new(OpenStackStfEventTestReceiver.new(build_qdr_client_url(options), DEFAULT_TOPIC_NAME))
+    qdr_client.run
+    true
+  rescue => ex
+    $log.info("OpenstackSTFEventMonitor availability check failed with #{ex}.") if $log
+    false
+  ensure
+    qdr_client.close if qdr_client.respond_to?(:close)
+    qdr_client.stop if qdr_client.respond_to?(:stop)
+  end
+
+  def self.plugin_priority
+    1
+  end
+
+  def self.build_qdr_client_url(options)
+    $log.debug("Building STF QDR client connection with #{options.inspect}..") if $log
+    protocol = options[:security_protocol].to_s.start_with?('ssl') ? 'amqps' : 'amqp'
+    hostname = options[:hostname]
+    port     = options[:port] || DEFAULT_AMQP_PORT
+    "#{protocol}://#{hostname}:#{port}"
+  end
+
+  def initialize(options = {})
+    $log.info("Building STF QDR client INIT #{options.inspect}..") if $log
+    @options = options
+    @ems = options[:ems]
+    @config = options.fetch(:stf, {})
+
+    @events = []
+    @events_mutex = Mutex.new
+
+    @recv_block = ->(event) { @events << event }
+
+    @qdr_receiver = Qpid::Proton::Container.new(OpenStackStfEventReceiver.new(self.class.build_qdr_client_url(@options), @config[:topic_name] || DEFAULT_TOPIC_NAME, @recv_block, @events_mutex))
+  end
+
+  def start
+    return if @qdr_receiver.running > 0
+
+    $log.info("STF QDR client START..") if $log
+    @handler = Thread.start { @qdr_receiver.run }
```

Comment:
Is there anything that would restart the handler if `@qdr_receiver` failed?  Thinking maybe `each_batch` can ensure the thread is running and restart it if not.

Also probably good to log any exceptions from the handler thread otherwise if it fails it will do so silently and just look like no events are coming through

Reply:
Good point! updating to raise failure when connection dies to let MIQ restart the Event monitor.

Reply:
Thanks for brining this up, this worked bad in previous solution. Updated to raise error which is logged and EventMonitor attempts to reconnect.

Reply:
:+1: nice